opengever/maintenance/scripts/validate_object_schema_conformance.py

```diff
@@ -0,0 +1,339 @@
+
+        # This is a dictionary with defaults to a depth of 3 meant
+        # for entries of the type stats[portal_type][field_name][error_type]
+        self.stats = defaultdict(lambda: defaultdict(Counter))
+        # stats["global"] keeps track of overall number of conforming objects
+        self.stats["global"] = Counter(conforming=0, non_conforming=0)
+        # stats["per_portal_type"] keeps track of number of conforming objects
+        # per portal type
+        self.stats["per_portal_type"] = defaultdict(Counter)
+
+        ts = datetime.today().strftime('%Y-%m-%d_%H-%M-%S')
+        self.csv_log_path = self.get_logfile_path(
+            'find-nonconforming-objects-%s.csv' % ts)
+        self.summary_log_path = self.get_logfile_path(
+            'find-nonconforming-objects-summary-%s.log' % ts)
+
+        if options.verbose:
+            self.error_repr = self.verbose_error_repr
+        else:
+            self.error_repr = self.non_verbose_error_repr
+
+    @staticmethod
+    def verbose_error_repr(err):
+        return repr(err)
+
+    @staticmethod
+    def non_verbose_error_repr(err):
+        return str(type(err))
+
+    def run(self):
+        sys.stderr.write("Checking for object not conforming to schema...\n\n")
+
+        all_brains = self.catalog.unrestrictedSearchResults()
+        total = len(all_brains)
+        object_generator = (brain.getObject() for brain in all_brains)
+
+        with open(self.csv_log_path, 'w') as self.csv_log:
+            with open(self.summary_log_path, 'w') as self.summary_log:
+                self.csv_log.write(self.CSV_HEADER + '\n')
+
+                for i, obj in enumerate(object_generator):
+                    missing_fields, invalid_fields, failed = self.validate_schema_conformance(obj)
+                    self.update_stats(obj, missing_fields, invalid_fields, failed)
+
+                    if missing_fields or invalid_fields or failed:
+                        self.write_csv_row(obj, missing_fields, invalid_fields, failed)
+
+                    if i % 100 == 0:
+                        sys.stderr.write("Progress: %s of %s objects\n" % (i, total))
+                self.display_stats()
+
+    def validate_schema_conformance(self, obj):
+        missing_fields = []
+        invalid_fields = []
+        failed_validations = []
+
+        portal_type = obj.portal_type
+        if portal_type not in self.SCHEMA_CACHE:
+            self.SCHEMA_CACHE[portal_type] = list(iterSchemataForType(portal_type))
+        schemas = self.SCHEMA_CACHE[portal_type]
+
+        for schema in schemas:
+            if schema.__identifier__ not in self.FIELD_CACHE:
+                self.FIELD_CACHE[schema.__identifier__] = getFieldsInOrder(schema)
+            fields = self.FIELD_CACHE[schema.__identifier__]
+
+            for name, field in fields:
+                field_instance = field.bind(obj)
+                try:
+                    value = field_instance.get(obj)
```

Comment:
This should be `field_instance.get(field.interface(obj))` (object needs to be adapted to schema to get a possible storage adapter). This might possibly explain some of the high nonconformance rates, especially for the LifeCycle behavior.

Reply:
Right, after correction, `LifeCycle` does not show up as much anymore, and overall now ~50% of the objects came out as conforming on my local installation (instead of the 10% measured before that fix).